src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -5280,6 +5280,55 @@ void C2_MacroAssembler::vector_mask_compress(KRegister dst, KRegister src, Regis
   kmov(dst, rtmp2);
 }
 
+#ifdef _LP64
+void C2_MacroAssembler::vector_compress_expand_avx2(int opcode, XMMRegister dst, XMMRegister src, XMMRegister mask,
+                                                    Register rtmp, Register rscratch, XMMRegister permv,
+                                                    XMMRegister xtmp, BasicType bt, int vec_enc) {
+  assert(type2aelembytes(bt) >= 4, "");
+  assert(opcode == Op_CompressV || opcode == Op_ExpandV, "");
+  if (bt == T_INT || bt == T_FLOAT) {
+    vmovmskps(rtmp, mask, vec_enc);
+    shlq(rtmp, 5);
+    if (opcode == Op_CompressV) {
+      lea(rscratch, ExternalAddress(StubRoutines::x86::compress_perm_table32()));
+    } else {
+      lea(rscratch, ExternalAddress(StubRoutines::x86::expand_perm_table32()));
+    }
+    addptr(rtmp, rscratch);
+    vmovdqu(permv, Address(rtmp));
+    vpermps(dst, permv, src, Assembler::AVX_256bit);
+    vpxor(xtmp, xtmp, xtmp, vec_enc);
+    // Blend the results with zero vector using permute vector as mask, its
+    // non-participating lanes holds a -1 value.
+    vblendvps(dst, dst, xtmp, permv, vec_enc);
```

Comment:
would you mind adding a few more comments to explain what happens here? I would also really appreciate more expressive register/variable names.

I think you are basically converting the `mask` to a permutation `permv`, by a lookup in the table.
Then you permute the `src` and blend it with a -1 vector, so that the unused (high) lanes are -1.

```
xtmp -> min_one
rtmp -> table_index
rscratch -> table_adr
```